test/dynamo/test_activation_checkpointing.py

```diff
@@ -1994,6 +1999,639 @@ def forward(self, primals_1: "f32[4, 4]"):
+    detach_recomputed = torch.ops.aten.detach.default(sigmoid_recomputed);  sigmoid_recomputed = None
+    detach_2 = torch.ops.aten.detach.default(detach_recomputed);  detach_recomputed = None
+    sigmoid_backward = torch.ops.aten.sigmoid_backward.default(expand, detach_2);  expand = detach_2 = None
+    t = torch.ops.aten.t.default(arg0_1);  arg0_1 = None
+    mm_2 = torch.ops.aten.mm.default(t, sigmoid_backward);  t = None
+    t_1 = torch.ops.aten.t.default(arg1_1);  arg1_1 = None
+    mm_3 = torch.ops.aten.mm.default(sigmoid_backward, t_1);  sigmoid_backward = t_1 = None
+    detach_3 = torch.ops.aten.detach.default(mm_3);  mm_3 = None
+    detach_4 = torch.ops.aten.detach.default(mm_2);  mm_2 = None
+    return (detach_3, detach_4)""",
+        )
+
+    @unittest.skipIf(not torch.cuda.is_available(), "CUDA not available")
+    def test_ac_rematerialize_defers_backward_only_nodes(self):
+        """AC nodes only used in backward are deferred (DCE removes forward version)."""
+
+        x_data = torch.randn(4, 4)
+
+        def forward_backward_with_ac():
+            x = x_data.detach().requires_grad_(True)
+            z = torch.utils.checkpoint.checkpoint(
+                lambda a: torch.sin(a), x, use_reentrant=False
+            )
+            loss = z.sum()
+            with torch.fx.traceback.annotate({"backward": 0}):
+                dx = _grad(loss, x)[0]
+            return dx.detach()
+
+        dx1, gm_without = self._compile_and_capture(forward_backward_with_ac, False)
+        dx2, gm_with = self._compile_and_capture(forward_backward_with_ac, True)
+
+        self.assertExpectedInline(
+            str(gm_without.code).strip(),
+            """\
+def forward(self, arg0_1):
+    sin = torch.ops.aten.sin.default(arg0_1)
+    sum_1 = torch.ops.aten.sum.default(sin);  sin = None
+    ones_like = torch.ops.aten.ones_like.default(sum_1, pin_memory = False, memory_format = torch.preserve_format);  sum_1 = None
+    expand = torch.ops.aten.expand.default(ones_like, [4, 4]);  ones_like = None
+    cos = torch.ops.aten.cos.default(arg0_1);  arg0_1 = None
+    mul = torch.ops.aten.mul.Tensor(expand, cos);  expand = cos = None
+    detach = torch.ops.aten.detach.default(mul);  mul = None
+    return (detach,)""",
+        )
+
+        self.assertExpectedInline(
+            str(gm_with.code).strip(),
+            """\
+def forward(self, arg0_1):
+    sin = torch.ops.aten.sin.default(arg0_1)
+    sum_1 = torch.ops.aten.sum.default(sin);  sin = None
+    ones_like = torch.ops.aten.ones_like.default(sum_1, pin_memory = False, memory_format = torch.preserve_format);  sum_1 = None
+    expand = torch.ops.aten.expand.default(ones_like, [4, 4]);  ones_like = None
+    cos = torch.ops.aten.cos.default(arg0_1);  arg0_1 = None
+    mul = torch.ops.aten.mul.Tensor(expand, cos);  expand = cos = None
+    detach = torch.ops.aten.detach.default(mul);  mul = None
+    return (detach,)""",
+        )
```

Comment:
These two graphs look identical, which is kind of expected because the output of sin isn't needed for backward, but it also seems that this test is not testing the deferred case, since nothing from forward is DCE'd.